addons/azureSets/azuresets.lua

```diff
@@ -75,41 +75,90 @@ function set_spells(spellset)
     log('Starting to set '..spellset..'.')
-    set_spells_from_spellset(spellset,1)
+    set_spells_from_spellset(spellset, 'remove')
     return
 end
 
-function set_spells_from_spellset(spellset,slot)
-    local islot
-    if tonumber(slot) < 10 then 
-        islot = '0'..slot
-    else islot = slot end
-    local tempname = settings.spellsets[spellset]['slot'..islot]
-    if tempname ~= nil then
-        for spell in spells:it() do
-                if spell['english']:lower() == tempname:lower() then
-                    windower.ffxi.set_blue_magic_spell(spell['id'], tonumber(slot))
-                    break
-                end
+function set_spells_from_spellset(spellset, setPhase)
+    local setToSet = settings.spellsets[spellset]
+    local currentSet = get_current_spellset()
+    --log_set(currentSet)
+    if setPhase == 'remove' then
+      -- Remove Phase
+      for k,v in pairs(currentSet) do
+        if not set_contains_value(setToSet, v) then
+          setSlot = k
+          local slotToRemove = tonumber(k:sub(5, k:len()))
+
+          windower.ffxi.remove_blue_magic_spell(slotToRemove)
+          --log('Removed spell: '..v..' at #'..slotToRemove)
+
+          windower.send_command('@wait .65;lua i azuresets set_spells_from_spellset '..spellset..' remove')
+          return
+        end
+      end
+    end
+    -- Did not find spell to remove. Start set phase
+    -- Find empty slot:
+    local slotToSetTo = -1
+    for i = 1, 20 do
+      local t = ''
+      if i < 10 then t = '0' end
+      local slotName = 'slot'..t..i
+      if currentSet[slotName] == nil then
+        slotToSetTo = i
+        break
+      end
+    end
+
+    if slotToSetTo ~= -1 then
+      -- We found an empty slot. Find a spell to set.
+      for k,v in pairs(setToSet) do
+        if not set_contains_value(currentSet, v) then
+          if v ~= nil then
+            local spellID = find_spell_id_by_name(v)
+            if spellID ~= -1 then
+              windower.ffxi.set_blue_magic_spell(spellID, tonumber(slotToSetTo))
+              --log('Set spell: '..v..' ('..spellID..') at: '..slotToSetTo)
+              windower.send_command('@wait .65;lua i azuresets set_spells_from_spellset '..spellset..' add')
+              return
+            end
+          end
         end
+      end
     end
-    if tonumber(slot) < 20 then
-        windower.send_command('@wait .5;lua i azuresets set_spells_from_spellset '..spellset..' '..slot+1)
-    else
-        log(spellset..' has been equipped.')
-        windower.send_command('@timers c "Blue Magic Cooldown" 60 up')
+
+    -- Unable to find any spells to set. Must be complete.
+    log(spellset..' has been equipped.')
+    windower.send_command('@timers c "Blue Magic Cooldown" 60 up')
+end
+
+function set_contains_value(set, value)
+  for k,v in pairs(set) do
+    if v:lower() == value:lower() then
+      return true
+    end
+  end
+  return false
+end
+
+function find_spell_id_by_name(spellname)
```

Comment:
This function can be removed. instead you can replace any call to this with:
```lua
spells:with('english', spellname:lower())
```

This will return the spell object or `nil`, if no object is found, then just use `.id` to get its ID.